filament/src/PostProcessManager.cpp

```diff
@@ -1893,38 +1931,73 @@ PostProcessManager::BloomPassOutput PostProcessManager::bloomPass(FrameGraph& fg
         bloomHeight *= inoutBloomOptions.anamorphism;
     }
 
-    // convert back to integer width/height
-    const uint32_t width  = std::max(1u, uint32_t(std::floor(bloomWidth)));
-    const uint32_t height = std::max(1u, uint32_t(std::floor(bloomHeight)));
-
     // we might need to adjust the max # of levels
     const uint32_t major = uint32_t(std::max(bloomWidth,  bloomHeight));
     const uint8_t maxLevels = FTexture::maxLevelCount(major);
     inoutBloomOptions.levels = std::min(inoutBloomOptions.levels, maxLevels);
     inoutBloomOptions.levels = std::min(inoutBloomOptions.levels, kMaxBloomLevels);
 
-    if (2 * width < desc.width || 2 * height < desc.height) {
-        // if we're scaling down by more than 2x, prescale the image with a blit to improve
-        // performance. This is important on mobile/tilers.
-        input = opaqueBlit(fg, input, { 0, 0, desc.width, desc.height }, {
-                .width = std::max(1u, desc.width / 2),
-                .height = std::max(1u, desc.height / 2),
-                .format = outFormat
-        });
+    if (inoutBloomOptions.quality == QualityLevel::LOW) {
+        // In low quality mode, we adjust the bloom buffer size so that both dimensions
+        // have enough exact mip levels. This can slightly affect the aspect ratio causing
+        // some artifacts:
+        // - add some anamorphism (experimentally not visible)
+        // - visible bloom size changes with dynamic resolution in non-homogenous mode
+        // This allows us to use the 9 sample downsampling filter (instead of 13)
+        // for at least 4 levels.
+        uint32_t width  = std::max(1u, uint32_t(std::floor(bloomWidth)));
+        uint32_t height = std::max(1u, uint32_t(std::floor(bloomHeight)));
+        width  &= ~((1 << 4) - 1);  // at least 4 levels
+        height &= ~((1 << 4) - 1);
+        bloomWidth  = float(width);
+        bloomHeight = float(height);
+    }
+
+    bool threshold = inoutBloomOptions.threshold;
+
+    while (2 * bloomWidth < float(desc.width) || 2 * bloomHeight < float(desc.height)) {
+        if (inoutBloomOptions.quality == QualityLevel::LOW ||
+            inoutBloomOptions.quality == QualityLevel::MEDIUM) {
+            input = downscalePass(fg, input, {
+                            .width  = (desc.width  = std::max(1u, desc.width  / 2)),
+                            .height = (desc.height = std::max(1u, desc.height / 2)),
+                            .format = outFormat
+                    },
+                    threshold, inoutBloomOptions.highlight, threshold);
```

Comment:
Did you intend to use `threshold` as the value for the `fireflies` param?

Reply:
yes, they're always happening together